src/features/Folding.ts

```diff
@@ -0,0 +1,602 @@
+                tokenStack.push(token);
+            }
+            if (token.scopes.indexOf(endScopeName) !== -1) {
+                result.unshift((new LineNumberRange(matchType)).fromTokenPair(tokenStack.pop(), token, document));
+            }
+        });
+
+        return result;
+    }
+
+    /**
+     * Given a textmate scope name, find a series of contiguous tokens which contain
+     * that scope name and pair them together.
+     * @param tokens    List of grammar tokens to parse
+     * @param scopeName The name of the scope to match
+     * @param matchType The type of range this region represents e.g. A comment
+     * @param document  The source text document
+     * @returns         A list of LineNumberRange objects of the contiguous token scopes
+     */
+    private matchContiguousScopeElements(
+        tokens: ITokenList,
+        scopeName: string,
+        matchType: vscode.FoldingRangeKind,
+        document: vscode.TextDocument,
+    ): ILineNumberRangeList {
+        const result = [];
+        let startToken;
+
+        tokens.forEach((token, index) => {
+            if (token.scopes.indexOf(scopeName) !== -1) {
+                if (startToken === undefined) { startToken = token; }
+
+                // If we're at the end of the token list, or the next token does not include the scopeName
+                // we've reached the end of the contiguous block.
+                if (((index + 1) >= tokens.length) || (tokens[index + 1].scopes.indexOf(scopeName) === -1)) {
+                    result.push((new LineNumberRange(matchType)).fromTokenPair(startToken, token, document));
+                    startToken = undefined;
+                }
+            }
+        });
+
+        return result;
+    }
+
+    /**
+     * Given a zero based offset, find the line text preceeding it in the document
+     * @param offset   Zero based offset in the document
+     * @param document The source text document
+     * @returns        The line text preceeding the offset, not including the preceeding Line Feed
+     */
+    private preceedingText(
+        offset: number,
+        document: vscode.TextDocument,
+    ): string {
+        const endPos = document.positionAt(offset);
+        const startPos = endPos.translate(0, -endPos.character);
+
+        return document.getText(new vscode.Range(startPos, endPos));
+    }
+
+    /**
+     * Given a zero based offset, find the line text after it in the document
+     * @param offset   Zero based offset in the document
+     * @param document The source text document
+     * @returns        The line text after the offset, not including the subsequent Line Feed
+     */
+    private subsequentText(
+        offset: number,
+        document: vscode.TextDocument,
+    ): string {
+        const startPos: vscode.Position = document.positionAt(offset);
+        const endPos: vscode.Position = document.lineAt(document.positionAt(offset)).range.end;
```

Comment:
I can't wait for JS to have the pipe operator :D 

https://github.com/tc39/proposal-pipeline-operator

Reply:
Always be trolling...